File: dda.py
import math
from pprint import pprint

class DDA:
  def __init__(self, cord):
    self.cord = cord

  def run(self):
    startX, startY, endX, endY = self.cord

    delX = endX - startX
    delY = endY - startY

    slope = delY / delX

    # step = None    
    # if slope <= 1: step = delX
    # else: step = delY
    
    step = delX if slope <=1 else delY

    xInc = delX / abs(step)
    yInc = delY / abs(step)

    output = [[startX, startY]]

    while True :
      if startX == endX or startY == endY: break

      startX += xInc
      startY += yInc

      output.append([startX, startY])

    return output
  
  
inputVal = list(map(int, input("Enter cordinates ").split()))

pprint(DDA(inputVal).run())

File: computer-graphics/dda.js
/* DDA (Digital Differential Algorithm) */

class DDA {
  constructor(cord1, cord2) {
    this.cord1 = cord1;
    this.cord2 = cord2;
  }

  run() {
    let [startX, startY] = this.cord1;
    let [endX, endY] = this.cord2;

    const delX = endX - startX;
    const delY = endY - startY;

    const slope = delY / delX;

    let step = slope <= 1 ? delX : delY;

    const xInc = delX / Math.abs(step);
    const yInc = delY / Math.abs(step);

    const output = [[startX, startY]];

    while (true) {
      if (startX === endX || startY === endY) break;

      startX += xInc;
      startY += yInc;

      output.push([startX, startY])
    }

    return output;
  }
}

const dda1 = new DDA([5, 4], [12, 7]).run()
const dda2 = new DDA([5, 7], [10, 15]).run()
const dda3 = new DDA([12, 9], [17, 14]).run()
const dda4 = new DDA([17, 14], [12, 9]).run()

console.log(dda1);
console.log(dda2);
console.log(dda3);
console.log(dda4);
